examples/context-propagation/kotlin/src/main/kotlin/example/armeria/contextpropagation/manual/MainService.kt

```diff
@@ -0,0 +1,104 @@
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.future.asDeferred
+import kotlinx.coroutines.future.await
+import kotlinx.coroutines.future.future
+import kotlinx.coroutines.withContext
+
+class MainService(private val backendClient: WebClient) : HttpService {
+    override fun serve(ctx: ServiceRequestContext, req: HttpRequest): HttpResponse {
+        val ctxExecutor = ctx.contextAwareExecutor()
+        val response = GlobalScope.future(ctxExecutor.asCoroutineDispatcher()) {
+            val numsFromRequest = async { fetchFromRequest(ctx, req) }
+            val numsFromDb = async { fetchFromFakeDb(ctx) }
+            val nums = awaitAll(numsFromRequest, numsFromDb).flatten()
+
+            val backendResponses =
+                awaitAll(
+                    *nums.map { num ->
+                        // The context is mounted in a thread-local, meaning it is available to all logic such
+                        // as tracing.
+                        require(ServiceRequestContext.current() === ctx)
+                        require(ctx.eventLoop().inEventLoop())
+                        backendClient.get("/square/$num").aggregate().asDeferred()
+                    }.toTypedArray()
+                ).toList()
+            // The context is mounted in a thread-local, meaning it is available to all logic such as tracing.
+            require(ServiceRequestContext.current() === ctx)
+            require(ctx.eventLoop().inEventLoop())
+            HttpResponse.of(
+                backendResponses.stream()
+                    .map(AggregatedHttpResponse::contentUtf8)
+                    .collect(Collectors.joining("\n"))
+            )
+        }
+        return HttpResponse.from(response)
+    }
+
+    private suspend fun fetchFromRequest(ctx: ServiceRequestContext, req: HttpRequest): List<Long> {
+        // The context is mounted in a thread-local, meaning it is available to all logic such as tracing.
+        require(ServiceRequestContext.current() === ctx)
+        require(ctx.eventLoop().inEventLoop())
+        val aggregatedHttpRequest = req.aggregate().await()
+        val nums = mutableListOf<Long>()
+        for (token in Iterables.concat(
+            NUM_SPLITTER.split(aggregatedHttpRequest.path().substring(1)),
+            NUM_SPLITTER.split(aggregatedHttpRequest.contentUtf8())
+        )) {
+            nums.add(token.toLong())
+        }
+        return nums
+    }
+
+    private suspend fun fetchFromFakeDb(ctx: ServiceRequestContext): List<Long> {
+        // The context is mounted in a thread-local, meaning it is available to all logic such as tracing.
+        require(ServiceRequestContext.current() === ctx)
+        require(ctx.eventLoop().inEventLoop())
+        // This logic mimics using a blocking method, which would usually be something like a MySQL
+        // database query using JDBC.
+        return withContext(ctx.blockingTaskExecutor().asCoroutineDispatcher()) {
```

Comment:
What do you think about `withContext` vs `CompletableFuture.supplyAsync().await()`? I'm a bit worried about using a suspending function inside this block which I think would change threads and possibly break something like a transaction. I guess suspension should generally be prevented for uses of blocking executor, do you know if there's a simpler way to do that?

Reply:
You mean user may do like below?
```
return withContext(ctx.blockingTaskExecutor().asCoroutineDispatcher()) {
            ..
            // begin tx
            ..
            anotherSuspendFunction();
            ..
            // end tx
        }
```

If so, maybe as you said we should wrap whole tx in one `CompletableFuture.supplyAsync()`.

Reply:
Yeah that's what I'm a bit worried about - maybe it's too much worrying ;) But I wish there were a good pattern to prevent suspension within a certain block of code, couldn't find anything with some googling but it seems important enough to be in Kotlin.

Reply:
Agree, anyway I have updated the code. I'm not familiar with reactive transaction https://spring.io/blog/2019/05/16/reactive-transactions-with-spring. Looks like it can allow us to suspend during tx.